sdks/python/apache_beam/runners/interactive/background_caching_job_test.py

```diff
@@ -99,6 +117,89 @@ def test_background_caching_job_not_start_when_such_job_is_done(self):
+
+  @patch('IPython.get_ipython', new_callable=mock_get_ipython)
+  def test_source_to_cache_changed_when_new_source_is_added(self, cell):
+    with cell:  # Cell 1
+      pipeline = _build_an_empty_stream_pipeline()
+      read_foo = pipeline | 'Read' >> beam.io.ReadFromPubSub(
+          subscription=_FOO_PUBSUB_SUB)
+      ib.watch({'read_foo': read_foo})
+
+    # Sets the signature for current pipeline state.
+    ie.current_env().set_cached_source_signature(
+        pipeline, bcj.extract_source_to_cache_signature(pipeline))
+
+    with cell:  # Cell 2
+      read_bar = pipeline | 'Read' >> beam.io.ReadFromPubSub(
+          subscription=_BAR_PUBSUB_SUB)
+      ib.watch({'read_bar': read_bar})
+
+    self.assertTrue(bcj.is_source_to_cache_changed(pipeline))
+
+  @patch('IPython.get_ipython', new_callable=mock_get_ipython)
+  def test_source_to_cache_changed_when_source_is_altered(self, cell):
+    with cell:  # Cell 1
+      pipeline = _build_an_empty_stream_pipeline()
+      transform = beam.io.ReadFromPubSub(
+          subscription=_FOO_PUBSUB_SUB)
+      read_foo = pipeline | 'Read' >> transform
+      ib.watch({'read_foo': read_foo})
+
+    # Sets the signature for current pipeline state.
+    ie.current_env().set_cached_source_signature(
+        pipeline, bcj.extract_source_to_cache_signature(pipeline))
+
+    with cell:  # Cell 2
+      from apache_beam.io.gcp.pubsub import _PubSubSource
+      # Alter the transform.
+      transform._source = _PubSubSource(subscription=_BAR_PUBSUB_SUB)
+
+    self.assertTrue(bcj.is_source_to_cache_changed(pipeline))
+
+  @patch('IPython.get_ipython', new_callable=mock_get_ipython)
+  def test_source_to_cache_not_changed_for_same_source(self, cell):
+    with cell:  # Cell 1
+      pipeline = _build_an_empty_stream_pipeline()
+      transform = beam.io.ReadFromPubSub(
+          subscription='projects/test-project/subscriptions/dummy')
+
+    with cell:  # Cell 2
+      read_foo_1 = pipeline | 'Read' >> transform
+      ib.watch({'read_foo_1': read_foo_1})
+
+    # Sets the signature for current pipeline state.
+    ie.current_env().set_cached_source_signature(
+        pipeline, bcj.extract_source_to_cache_signature(pipeline))
+
+    with cell:  # Cell 3
+      # Apply exactly the same transform and the same instance.
+      read_foo_2 = pipeline | 'Read' >> transform
+      ib.watch({'read_foo_2': read_foo_2})
+
+    self.assertFalse(bcj.is_source_to_cache_changed(pipeline))
+
+    with cell:  # Cell 4
+      # Apply the same transform but represented in a different instance.
+      # The signature representing the urn and payload is still the same, so it
+      # is not treated as a new unbounded source.
+      read_foo_3 = pipeline | 'Read' >> beam.io.ReadFromPubSub(
+          subscription='projects/test-project/subscriptions/dummy')
+      ib.watch({'read_foo_3': read_foo_3})
+
+    self.assertFalse(bcj.is_source_to_cache_changed(pipeline))
+
```

Comment:
Would it be a valid test case to remove a source and observe that cache is not changed? (Since the test is based on current signature being a subset of the previous signature.)

Reply:
Yes, added a test for the scenario.